src/colmap/estimators/cost_functions.h

```diff
@@ -47,27 +47,84 @@ template <typename T>
 using EigenQuaternionMap = Eigen::Map<const Eigen::Quaternion<T>>;
 using EigenMatrix6d = Eigen::Matrix<double, 6, 6>;
 
+enum class CovarianceType {
+  IDENTITY = 1,
+  DIAGONAL = 2,
+  DENSE = 3,
+};
+
+template <CovarianceType CTYPE = CovarianceType::DENSE>
 inline Eigen::MatrixXd SqrtInformation(const Eigen::MatrixXd& covariance) {
-  return covariance.inverse().llt().matrixL().transpose();
+  if constexpr (CTYPE == CovarianceType::IDENTITY) {
+    return covariance;
+  } else {
+    return covariance.inverse().llt().matrixL().transpose();
+  }
+}
+
+inline CovarianceType GetCovarianceType(const Eigen::MatrixXd& covariance) {
+  if (covariance.isDiagonal()) {
+    if (covariance.isIdentity()) {
+      return CovarianceType::IDENTITY;
+    } else {
+      return CovarianceType::DIAGONAL;
+    }
+  } else {
+    return CovarianceType::DENSE;
+  }
+}
+
+template <CovarianceType CTYPE>
+inline bool CheckCovarianceByType(const Eigen::MatrixXd& covariance) {
+  if constexpr (CTYPE == CovarianceType::IDENTITY) {
+    return covariance.isIdentity();
+  } else if constexpr (CTYPE == CovarianceType::DIAGONAL) {
+    return covariance.isDiagonal();
+  } else {
+    return true;
+  }
+}
+
+template <typename T, int N, CovarianceType CTYPE = CovarianceType::IDENTITY>
+inline void ApplySqrtInformation(
+    T* residuals, const Eigen::Matrix<double, N, N>& sqrt_information) {
+  if constexpr (CTYPE == CovarianceType::IDENTITY) {
+    return;
+  } else if constexpr (CTYPE == CovarianceType::DIAGONAL) {
+    for (int i = 0; i < sqrt_information.rows(); ++i) {
+      residuals[i] *= sqrt_information(i, i);
+    }
+  } else if constexpr (CTYPE == CovarianceType::DENSE) {
+    Eigen::Map<Eigen::Matrix<T, N, 1>> residuals_map(residuals);
+    residuals_map.applyOnTheLeft(sqrt_information.template cast<T>());
+  }
 }
 
 // Standard bundle adjustment cost function for variable
 // camera pose, calibration, and point parameters.
-template <typename CameraModel>
+template <typename CameraModel, CovarianceType CTYPE = CovarianceType::IDENTITY>
 class ReprojErrorCostFunction {
  public:
-  explicit ReprojErrorCostFunction(const Eigen::Vector2d& point2D)
-      : observed_x_(point2D(0)), observed_y_(point2D(1)) {}
+  explicit ReprojErrorCostFunction(
```

Comment:
no explicit

Reply:
The covariance argument is optional, and the compiler does not like it when i remove the "explicit": https://github.com/colmap/colmap/actions/runs/11029101065/job/30630763773